### PR TITLE
chore(l2): forbid index slicing

### LINKED ISSUE #1270: L2: Forbid indexing slicing
Add the following clippy lint to all `Cargo.toml` files in the `l2` crate:

```toml
[lints.clippy]
indexing_slicing = "deny"
```

and make the indexing safe where necessary.

### PR DESCRIPTION
**Pull request by dsocolobsky**
**Motivation**

We should not use direct indexing `var[idx]` or direct index slicing `var[0..2]` and instead use safe alternatives and handle the errors.

**Description**

- `indexing_slicing = "deny"` was added to `Cargo.toml` to forbid index slicing usage.
- All usages were removed in the `l2` crate

Closes #1270

### CODE REVIEW

File: crates/l2/proposer/l1_watcher.rs [hunk 1]
```diff
@@ -112,22 +115,28 @@ impl L1Watcher {
...
+                    self.last_block_fetched + 1,
+                    new_last_block,
+                    self.address,
+                    *first_topic,
+                )
+                .await
+            {
+                Ok(logs) => logs,
+                Err(error) => {
+                    warn!("Error when getting logs from L1: {}", error);
```
Comment (ilitteri) [inline crates/l2/proposer/l1_watcher.rs:132]:
```suggestion
                    warn!("Error when getting logs from L1: {error}");
```

File: crates/l2/proposer/l1_watcher.rs [hunk 2]
```diff
@@ -158,14 +167,31 @@ impl L1Watcher {
...
+            })?;
+            let beneficiary_uint = log
+                .log
+                .topics
+                .get(2)
+                .ok_or(L1WatcherError::FailedToDeserializeLog(
+                    "Failed to parse beneficiary from log: log.topics[2] out of bounds".to_owned(),
+                ))?
+                .into_uint();
+            let beneficiary = format!("{:#x}", beneficiary_uint)
```
Comment (ilitteri) [inline crates/l2/proposer/l1_watcher.rs:194]:
```suggestion
            let beneficiary = format!("{beneficiary_uint:#x}")
```

File: crates/l2/utils/eth_client/mod.rs [hunk 3]
```diff
@@ -365,7 +365,9 @@ impl EthClient {
 
         // Add the nonce just if present, otherwise the RPC will use the latest nonce
         if let Some(nonce) = transaction.nonce {
-            data["nonce"] = json!(format!("{:#x}", nonce));
+            if let Value::Object(ref mut map) = data {
+                map.insert("nonce".to_owned(), json!(format!("{:#x}", nonce)));
```
Comment (ilitteri) [inline crates/l2/utils/eth_client/mod.rs:369]:
```suggestion
                map.insert("nonce".to_owned(), json!(format!("{nonce:#x}")));
```

File: crates/l2/proposer/errors.rs [hunk 4]
```diff
@@ -92,6 +93,10 @@ pub enum CommitterError {
     FailedToReExecuteBlock(#[from] EvmError),
     #[error("Committer failed to send transaction: {0}")]
     FailedToSendCommitment(String),
+    #[error("Commiter failed to decode deposit hash")]
+    FailedToDecodeDepositHash,
+    #[error("Commiter failed to merkelize: {0}")]
```
Comment (fborello-lambda) [inline crates/l2/proposer/errors.rs:98]: nit: 
```suggestion
    #[error("Committer failed to decode deposit hash")]
    FailedToDecodeDepositHash,
    #[error("Committer failed to merkelize: {0}")]
```

**Review by ilitteri — COMMENTED** [review]
LGTM. Left some nits.

File: crates/l2/utils/mod.rs [hunk 5]
```diff
@@ -26,14 +25,23 @@ where
...
 }
 
-pub fn get_address_from_secret_key(secret_key: &SecretKey) -> Result<Address, TryFromSliceError> {
+pub fn get_address_from_secret_key(secret_key: &SecretKey) -> Result<Address, EthClientError> {
     let public_key = secret_key
         .public_key(secp256k1::SECP256K1)
         .serialize_uncompressed();
     let hash = keccak(&public_key[1..]);
 
     // Get the lat 20 bytes of the hash
```
Comment (fborello-lambda) [inline crates/l2/utils/mod.rs:34]: nit: 
```suggestion
    // Get the last 20 bytes of the hash
```

**Review by fborello-lambda — COMMENTED** [review]
looks good, left 2 suggestions

File: crates/l2/proposer/l1_watcher.rs (hunk 1 above)
Comment (dsocolobsky) [reply crates/l2/proposer/l1_watcher.rs]: fixed in bec14139b55734f5f34037cefff5ff577c151829

File: crates/l2/proposer/l1_watcher.rs (hunk 2 above)
Comment (dsocolobsky) [reply crates/l2/proposer/l1_watcher.rs]: fixed in bec14139b55734f5f34037cefff5ff577c151829

File: crates/l2/utils/eth_client/mod.rs (hunk 3 above)
Comment (dsocolobsky) [reply crates/l2/utils/eth_client/mod.rs]: fixed in bec14139b55734f5f34037cefff5ff577c151829

File: crates/l2/proposer/errors.rs (hunk 4 above)
Comment (dsocolobsky) [reply crates/l2/proposer/errors.rs]: fixed in bec14139b55734f5f34037cefff5ff577c151829

File: crates/l2/utils/mod.rs (hunk 5 above)
Comment (dsocolobsky) [reply crates/l2/utils/mod.rs]: fixed in bec14139b55734f5f34037cefff5ff577c151829

**Review by ilitteri — APPROVED** [review]
LGTM

**Review by fborello-lambda — APPROVED** [review]

**Pull request merged**